posts/2018-12-11-hunt-bugs-down-before-they-are-merged.md

```diff
@@ -0,0 +1,362 @@
+
+* When someone hits a bug in the main code base, they were most likely
+  not looking for one, but trying to get work done.
+* Whomever is tasked with fixing the bug may have lost (or never have had)
+  context relevant to the faulty code.
+* Reporting a bug creates noise in the form of tickets and pull requests (PRs).
+
+Let's jump right in!
+
+<div class="story">
+You are a frontend engineer for a company that stores online notes. You're
+tasked with writing an autocomplete function allowing the user to find notes
+more quickly:
+
+```
+
++-----------------------+  +------+
+| Gr|                   |  | open |
++-----------------------+  +------+
+|                       |
+| Great minds.txt       |
+|                       |
+| Grocery list          |
+|                       |
+| Grrr! said the lion   |
+|                       |
+| ...                   |
++-----------------------+
+
+```
+
+The following HTTP call returns the list of all of a user's notes:
+
+``` http
+GET /users/<user>/notes?query=<prefix>
+```
+</div>
+
+### It gets in the way
+
+<div class="story" style="display: inline-block;">
+
+<figure style="float: left;">
+<img src="../images/peanut_butter_cookies_nola.jpg" alt="Rated #1 in NOLA"><figcaption>And #2 in America!</figcaption>
+</figure>
+
+You've been doing amazing work. You got the flow. Mind like water. You're done
+with your frontend changes. You spin up a local instance of the server.
+
+You try it out and realize that the server segfaults when there are no matches.
+
+You despair. There goes your flow. The name of the backend engineers flash
+before your eyes. You consider not inviting them to your wedding. You make a
+mental list of who's most likely to have introduced the bug. You consider
+quitting your job and becoming a barista in a different country -- maybe even
+in New Orleans.  You'll need a specialty, maybe peanut butter cookies. You'll
+kill it with your peanut butter cookies. People will know about them from New
+England to New Mexico. You'll become famous and, in a few years, run for
+president of the United States. Then you consider the logistics: your fiancée
+will need to quit her job, you'll need to hire a moving service, your dog may
+not survive the flight.
+
+You stash your local changes and set out to write a test case reproducing the
+issue.
+</div>
+
+Unless you're a quality analyst, you probably don't go around trying to find
+bugs for fun. Most of us discover bugs while trying to get something done:
+maybe you're simply using the server, or you're writing a new feature that
+depends on a different feature or function. Discovering wrong behavior in a
+codebase is almost always more of an annoyance than a nice surprise.
+
+Had the bug been caught before merge, the bug reporter would most likely have
+been the person who introduced the faulty behavior, and would have most likely
+been expecting to encounter "some" bug (correct software on the first attempt
+is a myth, sorry).
```

Comment:
after reading part I think the intro is not exactly precise, since this really more about exposing a bug that was already there rather than introducing a new one.  I think the intro should be updated so it includes this explicitly?